File: _stats_build/bin/gen_wr_ids.rb
# NOTE: 从 RoundMetric 的 ranking_data 提取每个项目的 WR 值和前 2 名 WCA ID
# 输出极简 JSON，前端通过 WCA ID 调 API 获取详细数据（times / ao100）
# 不依赖 AverageOfX，不需要 Ao100 查询
#
# 用法：cd _stats_build && ruby bin/gen_wr_ids.rb

ENV["STATS_USE_CACHE"] = "1"

require_relative "../statistics/index"
require_relative "../core/events"
require "json"

OUTPUT_PATH = File.expand_path("../../stats/wr_ids.json", __dir__)

result = {}

# NOTE: 从 person_link markdown 提取 WCA ID
# 格式: [Name](https://www.worldcubeassociation.org/persons/2023GENG02)
def extract_wca_id(person_link)
  person_link[%r{/persons/([^)\]]+)}, 1]
end

# NOTE: 从 metric_str 解析为 centiseconds
def parse_cs(str)
  case str
  when /\A(\d+):(\d+\.\d+)\z/
    $1.to_i * 6000 + ($2.to_f * 100).round
  when /\A\d+\.\d+\z/
    (str.to_f * 100).round
  when /\A\d+\z/
    str.to_i * 100
  end
end

# NOTE: 提取 average WR top 2 的 WCA ID + 成绩
avg_klass = WrAverageHistory.new
avg_klass.ranking_data.each do |event_name, top10|
  next if top10.empty?

  event_id = Events::ALL.find { |id, name| name == event_name }&.first
  next unless event_id

  result[event_id] ||= {}

  top10.first(2).each_with_index do |row, idx|
    # row 格式: [rank, person_link, metric_str, country, date, comp_link]
    wca_id = extract_wca_id(row[1])
    cs = parse_cs(row[2])
    next unless wca_id && cs && cs > 0

    suffix = idx == 0 ? "1" : "2"
    result[event_id]["avg_id_#{suffix}"] = wca_id
    result[event_id]["avg_#{suffix}"] = cs
  end
end

# NOTE: 提取 single WR
single_klass = WrSingleHistory.new
single_klass.ranking_data.each do |event_name, top10|
  next if top10.empty?

  event_id = Events::ALL.find { |id, name| name == event_name }&.first
  next unless event_id

  result[event_id] ||= {}
  cs = parse_cs(top10.first[2])
  result[event_id]["single"] = cs if cs && cs > 0
end

# NOTE: 校验
has_ids = result.any? { |_, v| v.key?("avg_id_1") }
unless has_ids
  abort "FATAL: wr_ids.json has no player IDs. Check WrAverageHistory."
end

File.write(OUTPUT_PATH, JSON.pretty_generate(result) + "\n")
puts "Generated #{OUTPUT_PATH} (#{result.size} events)"

File: core/jobs/stats-build/src/bin/gen_wr_ids.ts
// NOTE: 从已生成的 wr_metric.json 中提取每个项目的 WR 值和 top 2 WCA ID
// 直接读取 compute_all 已生成的 JSON，无需 MySQL 连接
//
// 输出：stats/wr_ids.json — calc 页面依赖此文件获取当前 WR 数据
// 用法：npx tsx src/bin/gen_wr_ids.ts
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_DIR = resolve(__dirname, '../../../../../stats');
const OUTPUT_PATH = resolve(__dirname, '../../../../../stats/wr_ids.json');

// NOTE: 项目中英文名 → 项目 ID 映射（从 events.ts 概念反转而来）
// 因为 wr_metric.json 的 sections 用英文名作 title
import { EVENTS } from '../core/events.js';

// NOTE: 反转映射：英文名 → 项目 ID
const NAME_TO_ID: Record<string, string> = {};
for (const [id, name] of Object.entries(EVENTS)) {
  NAME_TO_ID[name] = id;
}

// NOTE: 从 person_link markdown 提取 WCA ID
// 格式: [Name](https://www.worldcubeassociation.org/persons/2023GENG02)
function extractWcaId(personLink: string): string | null {
  const match = personLink.match(/\/persons\/([^)\]]+)/);
  return match ? match[1] : null;
}

// NOTE: 从显示字符串解析为 centiseconds
function parseCentiseconds(str: string): number | null {
  // 格式: "1:23.45" → 8345
  const mmss = str.match(/^(\d+):(\d+\.\d+)$/);
  if (mmss) {
    return Math.round(Number(mmss[1]) * 6000 + Number(mmss[2]) * 100);
  }
  // 格式: "6.78" → 678
  const sec = str.match(/^\d+\.\d+$/);
  if (sec) {
    return Math.round(Number(str) * 100);
  }
  // 格式: "23" (FMC moves) → 2300
  const int = str.match(/^\d+$/);
  if (int) {
    return Number(str) * 100;
  }
  return null;
}

interface WrIdsEntry {
  avg_id_1?: string;
  avg_1?: number;
  avg_id_2?: string;
  avg_2?: number;
  single?: number;
}

function main() {
  const wrMetricPath = resolve(DATA_DIR, 'wr_metric.json');
  if (!existsSync(wrMetricPath)) {
    console.error(`FATAL: ${wrMetricPath} not found. Run compute_all.ts first.`);
    process.exit(1);
  }

  const wrMetric = JSON.parse(readFileSync(wrMetricPath, 'utf-8'));
  const result: Record<string, WrIdsEntry> = {};

  // NOTE: 提取 average WR top 2 — 从 metricPanels 中 id='average' 的 ranking panel
  const avgPanel = wrMetric.metricPanels?.find((mp: Record<string, unknown>) => mp.id === 'average');
  if (avgPanel) {
    const rankingPanel = avgPanel.panels?.find((p: Record<string, unknown>) => p.id === 'ranking');
    if (rankingPanel) {
      for (const section of rankingPanel.sections) {
        const eventId = NAME_TO_ID[section.title];
        if (!eventId) continue;

        if (!result[eventId]) result[eventId] = {};

        // NOTE: ranking rows 格式: [rank, person_link, result, country, date, comp_link, details]
        const top2 = (section.rows as unknown[][]).slice(0, 2);
        for (let idx = 0; idx < top2.length; idx++) {
          const row = top2[idx];
          const personLink = String(row[1]);
          const resultStr = String(row[2]);
          const wcaId = extractWcaId(personLink);
          const cs = parseCentiseconds(resultStr);
          if (!wcaId || !cs || cs <= 0) continue;

          const suffix = idx === 0 ? '1' : '2';
          result[eventId][`avg_id_${suffix}` as keyof WrIdsEntry] = wcaId as never;
          result[eventId][`avg_${suffix}` as keyof WrIdsEntry] = cs as never;
        }
      }
    }
  }

  // NOTE: 提取 single WR — 从 metricPanels 中 id='single' 的 ranking panel
  const singlePanel = wrMetric.metricPanels?.find((mp: Record<string, unknown>) => mp.id === 'single');
  if (singlePanel) {
    const rankingPanel = singlePanel.panels?.find((p: Record<string, unknown>) => p.id === 'ranking');
    if (rankingPanel) {
      for (const section of rankingPanel.sections) {
        const eventId = NAME_TO_ID[section.title];
        if (!eventId) continue;

        if (!result[eventId]) result[eventId] = {};

        const firstRow = section.rows?.[0] as unknown[] | undefined;
        if (!firstRow) continue;

        const cs = parseCentiseconds(String(firstRow[2]));
        if (cs && cs > 0) result[eventId].single = cs;
      }
    }
  }

  // NOTE: 校验
  const hasIds = Object.values(result).some(v => v.avg_id_1);
  if (!hasIds) {
    console.error('FATAL: wr_ids.json has no player IDs. Check wr_metric.json.');
    process.exit(1);
  }

  writeFileSync(OUTPUT_PATH, JSON.stringify(result, null, 2) + '\n', 'utf-8');
  console.log(`Generated ${OUTPUT_PATH} (${Object.keys(result).length} events)`);
}

main();
